power_grid_model_c/power_grid_model/include/power_grid_model/main_core/update.hpp

```diff
@@ -130,4 +135,194 @@ inline void update_inverse(MainModelState<ComponentContainer> const& state, Forw
+    properties.elements_ps_in_update =
+        properties.uniform ? update_data.uniform_elements_per_scenario(properties.name) : invalid_index;
+    properties.elements_in_base = n_component;
+
+    if (properties.is_columnar) {
+        detail::process_buffer_span<CompType>(
+            update_data.template get_columnar_buffer_span_all_scenarios<meta_data::update_getter_s, CompType>(),
+            properties);
+    } else {
+        detail::process_buffer_span<CompType>(
+            update_data.template get_buffer_span_all_scenarios<meta_data::update_getter_s, CompType>(), properties);
+    }
+
+    return properties;
+}
+
+inline void validate_update_data_independence(detail::UpdateCompProperties const& comp) {
+    if (comp.is_empty_component()) {
+        return; // empty dataset is still supported
+    }
+    auto const elements_ps = comp.get_n_elements();
+    assert(comp.uniform || elements_ps < 0);
+
+    if (elements_ps >= 0 && comp.elements_in_base < elements_ps) {
+        throw DatasetError("Update data has more elements per scenario than input data for component " + comp.name +
+                           "!");
+    }
+    if (comp.ids_part_na) {
+        throw DatasetError("Some IDs are not valid for component " + comp.name + " in update data!");
+    }
+    if (comp.ids_all_na && comp.elements_in_base != elements_ps) {
+        throw DatasetError("Update data without IDs for component " + comp.name +
+                           " has a different number of elements per scenario then input data!");
+    }
+}
+
+} // namespace detail
+
+// main core utils
+template <class CompType, class... ComponentTypes>
+constexpr auto index_of_component = main_core::utils::index_of_component<CompType, ComponentTypes...>;
+template <class... ComponentTypes> using SequenceIdx = main_core::utils::SequenceIdx<ComponentTypes...>;
+template <class... ComponentTypes> using ComponentFlags = main_core::utils::ComponentFlags<ComponentTypes...>;
+
+// get sequence idx map of a certain batch scenario
+template <typename CompType, class ComponentContainer>
+std::vector<Idx2D> get_component_sequence(MainModelState<ComponentContainer> const& state,
+                                          ConstDataset const& update_data, Idx scenario_idx,
+                                          detail::UpdateCompProperties const& comp_independence = {}) {
+    auto const get_sequence = [&state, n_comp_elements = comp_independence.get_n_elements()](auto const& span) {
+        return get_component_sequence<CompType>(state, std::begin(span), std::end(span), n_comp_elements);
+    };
+    if (update_data.is_columnar(CompType::name)) {
+        auto const buffer_span =
+            update_data.get_columnar_buffer_span<meta_data::update_getter_s, CompType>(scenario_idx);
+        return get_sequence(buffer_span);
+    }
+    auto const buffer_span = update_data.get_buffer_span<meta_data::update_getter_s, CompType>(scenario_idx);
+    return get_sequence(buffer_span);
+}
```

Comment:
I wonder if we can move the if columnar - else logic to dataset as `get_row_or_columnar buffer_span` since we use it everywhere. 
Maybe this is asked before but refresh my memory please.

Reply:
Then we might as well call it direct `get_buffer_span` ;)

Reply:
I already tried it when implementing `get_columnar_buffer_span` a few months ago, as I thought it would be more neat, but failed after several attempts. The biggest problem is that in C++ we can't have two different return types for the same overload. To solve this, one can use `std::any` or `std::variant`, but those come with additional overhead, which we want to avoid at this low level. Hence I was suggested to just use two separate overloads.

Maybe some sort of template shenanigan could be implemented, but my feeling is that the change would be not only difficult but may involve changing several other things. So, we could open some discussion about it if needed, but definitely scope it out of this PR.

Reply:
100% agree with this.

a potential solution would probably look similar to the component type functor application, but, for now, it's wayyyyyy overkill